File: pypi/benches/two_throughput.py
"""Ashwa search_two Throughput & Latency Microbenchmark Suite (Python / PyO3 Native C-Extension)"""

import os
import sys
import time

# Ensure package directory is on sys.path if running directly
_PKG_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "python"))
if _PKG_DIR not in sys.path:
    sys.path.insert(0, _PKG_DIR)

import ashwa

KB = 0x400
MB = KB * KB
GB = 0x400 * MB
SAMPLES = 0x200

TIERS = [
    {"name": "L1", "size": 0x20 * KB},
    {"name": "L2", "size": 0x200 * KB},
    {"name": "L3", "size": 0x10 * MB},
    {"name": "RAM", "size": 0x100 * MB},
    {"name": "RAM", "size": 0x200 * MB},
    {"name": "RAM", "size": 1 * GB},
]

# Anti-DCE (Dead Code Elimination) optimization barrier sink
black_hole = None

def format_size(bytes_count: int) -> str:
    if bytes_count >= GB:
        return f"{bytes_count // GB} GiB"
    elif bytes_count >= MB:
        return f"{bytes_count // MB} MiB"
    elif bytes_count >= KB:
        return f"{bytes_count // KB} KiB"
    return f"{bytes_count} B"


def format_latency(secs: float) -> str:
    nanos = secs * 1e9
    if nanos < 1_000.0:
        return f"{nanos:.2f} ns"
    elif nanos < 1_000_000.0:
        return f"{nanos / 1_000.0:.2f} µs"
    elif nanos < 1_000_000_000.0:
        return f"{nanos / 1_000_000.0:.2f} ms"
    return f"{secs:.2f} s"

def benchmark_tier(tier: dict, haystack: bytearray, needle: bytes) -> dict:
    global black_hole
    size = tier["size"]
    view = memoryview(haystack)[:size]
    search_two = ashwa.search_two

    # Warmup phase
    warmup_start = time.perf_counter_ns()
    warmup_iters = 0

    while (
        (time.perf_counter_ns() - warmup_start) < 0x5F5E100  # 100ms
        and warmup_iters < 0x40
    ) or warmup_iters < 2:
        black_hole = search_two(view, needle)
        warmup_iters += 1

    probe_start = time.perf_counter_ns()
    probe_iters = max(0x0A, warmup_iters // 0x0A)

    for _ in range(probe_iters):
        black_hole = search_two(view, needle)

    probe_elapsed_secs = (time.perf_counter_ns() - probe_start) / 1e9
    time_per_single_iter = max(probe_elapsed_secs / probe_iters, 1e-9)

    batch_size = max(1, round(0.001 / time_per_single_iter))
    sample_durations = [0.0] * SAMPLES

    for s in range(SAMPLES):
        sample_start = time.perf_counter_ns()

        for _ in range(batch_size):
            black_hole = search_two(view, needle)

        elapsed_secs = (time.perf_counter_ns() - sample_start) / 1e9
        sample_durations[s] = elapsed_secs / batch_size

    sample_durations.sort()
    median_secs = sample_durations[len(sample_durations) // 2]
    gib_per_sec = (size / (1024.0 * 1024.0 * 1024.0)) / median_secs

    return {
        "name": tier["name"],
        "size": size,
        "latency_secs": median_secs,
        "throughput_gib": gib_per_sec,
    }

def print_table(results: list) -> None:
    col_tier = "Tier / Level"
    col_size = "Size"
    col_lat = "Latency (Median)"
    col_thrpt = "Throughput"

    w_tier = 0x16  
    w_size = 0x0A   
    w_lat = 0x12   
    w_thrpt = 0x10 

    divider = (
        f"+-{'-' * w_tier}-+-{'-' * w_size}-+-{'-' * w_lat}-+-{'-' * w_thrpt}-+"
    )

    print(divider)
    print(
        f"| {col_tier.ljust(w_tier)} | {col_size.rjust(w_size)} | {col_lat.rjust(w_lat)} | {col_thrpt.rjust(w_thrpt)} |"
    )
    print(divider)

    for r in results:
        size_str = format_size(r["size"])
        lat_str = format_latency(r["latency_secs"])
        thrpt_str = f"{r['throughput_gib']:.2f} GiB/s"
        print(
            f"| {r['name'].ljust(w_tier)} | {size_str.rjust(w_size)} | {lat_str.rjust(w_lat)} | {thrpt_str.rjust(w_thrpt)} |"
        )

    print(divider)

def main() -> None:
    needle = b"\x0a\x0b"
    max_size = max(t["size"] for t in TIERS)

    haystack = bytearray(max_size)

    for page_offset in range(0, max_size, 0x1000):  
        haystack[page_offset] = 0

    results = []
    for tier in TIERS:
        results.append(benchmark_tier(tier, haystack, needle))

    print_table(results)

if __name__ == "__main__":
    main()

File: core/benches/two_throughput.rs
use ashwa::search_two;
use std::alloc::{alloc_zeroed, dealloc, Layout};
use std::hint::black_box;
use std::time::{Duration, Instant};

const KB: usize = 0x400;
const MB: usize = KB * KB;
const GB: usize = 0x400 * MB;
const SAMPLES: usize = 0x200;

struct TierConfig {
    name: &'static str,
    size: usize,
}

const TIERS: [TierConfig; 6] = [
    TierConfig { name: "L1", size: 0x20 * KB },
    TierConfig { name: "L2", size: 0x200 * KB },
    TierConfig { name: "L3", size: 0x10 * MB },
    TierConfig { name: "RAM", size: 0x100 * MB },
    TierConfig { name: "RAM", size: 0x200 * MB },
    TierConfig { name: "RAM", size: 1 * GB },
];

struct BenchResult {
    name: &'static str,
    size: usize,
    latency_secs: f64,
    throughput_gib: f64,
}

fn format_size(bytes: usize) -> String {
    if bytes >= GB {
        format!("{} GiB", bytes / GB)
    } else if bytes >= MB {
        format!("{} MiB", bytes / MB)
    } else if bytes >= KB {
        format!("{} KiB", bytes / KB)
    } else {
        format!("{} B", bytes)
    }
}

fn format_latency(secs: f64) -> String {
    let nanos = secs * 1e9;

    if nanos < 1_000.0 {
        format!("{:.2} ns", nanos)
    } else if nanos < 1_000_000.0 {
        format!("{:.2} µs", nanos / 1_000.0)
    } else if nanos < 1_000_000_000.0 {
        format!("{:.2} ms", nanos / 1_000_000.0)
    } else {
        format!("{:.2} s", secs)
    }
}

fn benchmark_tier(tier: &TierConfig, haystack: &[u8], needle: [u8; 2]) -> BenchResult {
    let size = tier.size;
    let slice = &haystack[..size];

    // warmup
    let warmup_start = Instant::now();
    let mut warmup_iters = 0usize;

    while (warmup_start.elapsed() < Duration::from_millis(0x64) && warmup_iters < 0x40)
        || warmup_iters < 2
    {
        black_box(search_two(black_box(slice), black_box(needle)));
        warmup_iters += 1;
    }

    let probe_start = Instant::now();
    let probe_iters = 0x0A.max(warmup_iters / 0x0A);

    for _ in 0..probe_iters {
        black_box(search_two(black_box(slice), black_box(needle)));
    }

    let probe_elapsed = probe_start.elapsed().as_secs_f64();
    let time_per_single_iter = (probe_elapsed / probe_iters as f64).max(1e-9);
    let batch_size = ((0.001 / time_per_single_iter).round() as usize).max(1);

    let mut sample_durations = Vec::with_capacity(SAMPLES);
    for _ in 0..SAMPLES {
        let sample_start = Instant::now();
        for _ in 0..batch_size {
            black_box(search_two(black_box(slice), black_box(needle)));
        }

        let elapsed = sample_start.elapsed();
        let per_iter_secs = elapsed.as_secs_f64() / (batch_size as f64);
        sample_durations.push(per_iter_secs);
    }

    sample_durations.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let median_secs = sample_durations[sample_durations.len() / 2];
    let gib_per_sec = (size as f64 / (1024.0 * 1024.0 * 1024.0)) / median_secs;

    BenchResult { name: tier.name, size, latency_secs: median_secs, throughput_gib: gib_per_sec }
}

fn print_table(results: &[BenchResult]) {
    let col_tier = "Tier / Level";
    let col_size = "Size";
    let col_lat = "Latency (Median)";
    let col_thrpt = "Throughput";

    let w_tier = 0x16;
    let w_size = 0x0A;
    let w_lat = 0x12;
    let w_thrpt = 0x10;

    let divider = format!(
        "+-{:-<w_tier$}-+-{:-<w_size$}-+-{:-<w_lat$}-+-{:-<w_thrpt$}-+",
        "",
        "",
        "",
        "",
        w_tier = w_tier,
        w_size = w_size,
        w_lat = w_lat,
        w_thrpt = w_thrpt
    );

    println!("{}", divider);
    println!(
        "| {:<w_tier$} | {:>w_size$} | {:>w_lat$} | {:>w_thrpt$} |",
        col_tier,
        col_size,
        col_lat,
        col_thrpt,
        w_tier = w_tier,
        w_size = w_size,
        w_lat = w_lat,
        w_thrpt = w_thrpt
    );
    println!("{}", divider);

    for r in results {
        println!(
            "| {:<w_tier$} | {:>w_size$} | {:>w_lat$} | {:>w_thrpt$} |",
            r.name,
            format_size(r.size),
            format_latency(r.latency_secs),
            format!("{:.2} GiB/s", r.throughput_gib),
            w_tier = w_tier,
            w_size = w_size,
            w_lat = w_lat,
            w_thrpt = w_thrpt
        );
    }

    println!("{}", divider);
}

fn main() {
    let needle = [0x0Au8, 0x0Bu8];
    let max_size = TIERS.iter().map(|t| t.size).max().unwrap_or(1 * GB);

    let layout = Layout::from_size_align(max_size, 0x40).expect("valid layout");
    let ptr = unsafe { alloc_zeroed(layout) };

    if ptr.is_null() {
        panic!("failed to allocate benchmark buffer");
    }

    for page_offset in (0..max_size).step_by(0x1000) {
        unsafe { std::ptr::write_volatile(ptr.add(page_offset), 0) };
    }

    let haystack = unsafe { std::slice::from_raw_parts(ptr, max_size) };
    let mut results = Vec::with_capacity(TIERS.len());

    for tier in &TIERS {
        let result = benchmark_tier(tier, haystack, needle);
        results.push(result);
    }

    print_table(&results);
    unsafe { dealloc(ptr, layout) };
}
